system/lib/wasmfs/backends/opfs_backend.cpp

```diff
@@ -101,92 +114,221 @@ namespace {
 
 using ProxyWorker = emscripten::ProxyWorker;
 
+class OpenState {
+public:
+  enum Kind { None, Access, Blob };
+
+private:
+  Kind kind = None;
+  int id = -1;
+  oflags_t openFlags;
+  size_t openCount = 0;
+
+public:
+  Kind getKind() { return kind; }
+
+  int open(ProxyWorker& proxy, int fileID, oflags_t flags) {
+    int result;
+    if (openCount == 0) {
+      assert(kind == None);
+      switch (flags) {
+        case O_RDWR:
+        case O_WRONLY:
+          // If we need write access, try to open an AccessHandle.
+          proxy(
+            [&](auto ctx) { _wasmfs_opfs_open_access(ctx.ctx, fileID, &id); });
+          // TODO: Fall back to open as a blob instead.
+          if (id < 0) {
+            return -EACCES;
+          }
+          kind = Access;
+          break;
+        case O_RDONLY:
+          proxy(
+            [&](auto ctx) { _wasmfs_opfs_open_blob(ctx.ctx, fileID, &id); });
+          if (id < 0) {
+            return -EACCES;
+          }
+          kind = Blob;
+          break;
+        default:
+          WASMFS_UNREACHABLE("Unexpected open access mode");
+      }
+    } else if (openFlags == O_RDONLY &&
+               (flags == O_WRONLY || flags == O_RDWR)) {
+      // Try to upgrade to an AccessHandle.
+      int newID;
+      proxy(
+        [&](auto ctx) { _wasmfs_opfs_open_access(ctx.ctx, fileID, &newID); });
+      if (newID < 0) {
+        return -EACCES;
+      }
+      // We have an AccessHandle, so close the blob.
+      proxy([&]() { _wasmfs_opfs_close_blob(getBlobID()); });
+      id = newID;
+      kind = Access;
+    } else {
+      // Reuse existing object.
+      ++openCount;
+      return 0;
```

Comment:
Does `openFlags` not need to be updated in this case? E.g. if this was opened for `WRONLY` and was now opened for `RDONLY` should the state not be `RDWR`?

Reply:
Ah, actually I noticed I can just get rid of `openFlags` entirely and use only `kind`...